nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java

```diff
@@ -166,317 +108,158 @@
-        }
-
-        while (!incomingMessages.isEmpty()) {
-            SmtpEvent message = incomingMessages.poll();
-
-            if (message == null) {
-                return;
-            }
-
-            synchronized (message) {
-                if (resultCodeSetAndIsError(message)) {
-                    SMTPResultCode resultCode = SMTPResultCode.fromCode(message.getReturnCode());
-                    getLogger().warn("Message failed before onTrigger processing message was: " + resultCode.getLogMessage());
-                    continue;
-                }
-
-                try {
-                    FlowFile flowfile = session.create();
-
-                    if (message.getMessageData() != null) {
-                        flowfile = session.write(flowfile, out -> {
-                            InputStream inputStream = message.getMessageData();
-                            byte [] buffer = new byte[1024];
-
-                            int rd;
-                            long totalBytesRead =0;
-
-                            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1 ) {
-                                totalBytesRead += rd;
-                                if (totalBytesRead > server.getMaxMessageSize() ) {
-                                    message.setReturnCode(500);
-                                    message.setProcessed();
-                                    break;
-                                }
-                                out.write(buffer, 0, rd);
-                            }
-                            out.flush();
-                        });
-                    } else {
-                        getLogger().debug("Message body was null");
-                        message.setReturnCode(SMTPResultCode.UNKNOWN_ERROR_CODE.getCode());
-                        message.setProcessed();
-                    }
-
-                    if (!message.getProcessed()) {
-                        HashMap<String, String> attributes = new HashMap<>();
-                        // Gather message attributes
-                        attributes.put(SMTP_HELO, message.getHelo());
-                        attributes.put(SMTP_SRC_IP, message.getHelo());
-                        attributes.put(SMTP_FROM, message.getFrom());
-                        attributes.put(SMTP_TO, message.getTo());
-
-                        List<Map<String, String>> details = message.getCertifcateDetails();
-                        int c = 0;
-
-                        // Add a selection of each X509 certificates to the already gathered attributes
-
-                        for (Map<String, String> detail : details) {
-                            attributes.put("smtp.certificate." + c + ".serial", detail.getOrDefault("SerialNumber", null));
-                            attributes.put("smtp.certificate." + c + ".subjectName", detail.getOrDefault("SubjectName", null));
```

Comment:
It boils down to chain of custody. 

Imagine Joe Blogs and Luo Cipher both work for ACME corp and are issued certificated to send messages to my NiFi instance using TLS and requiring client Auth.

Since both are authorised, you want to distinguish from your authorised users who sent a particular email to NiFi, hence you store the subjectName used to send the message.

This by the way is the same approach used around NiFi

![image](https://cloud.githubusercontent.com/assets/3108527/17577508/ec07a332-5fc1-11e6-8d1a-50bb29edc9e2.png)

https://nifi.apache.org/docs/nifi-docs/components/org.apache.nifi.processors.standard.HandleHttpRequest/index.html